src/compiler/utilities.ts

```diff
@@ -1811,19 +1814,70 @@ namespace ts {
         };
     }
 
+    export function getExtensionsToRemoveForEmitPath(compilerOptons: CompilerOptions) {
+        return getSupportedExtensions(compilerOptons).concat("jsx", "js");
+    }
+
     export function getOwnEmitOutputFilePath(sourceFile: SourceFile, host: EmitHost, extension: string) {
         let compilerOptions = host.getCompilerOptions();
         let emitOutputFilePathWithoutExtension: string;
         if (compilerOptions.outDir) {
-            emitOutputFilePathWithoutExtension = removeFileExtension(getSourceFilePathInNewDir(sourceFile, host, compilerOptions.outDir));
+            emitOutputFilePathWithoutExtension = removeFileExtension(getSourceFilePathInNewDir(sourceFile, host, compilerOptions.outDir),
+                getExtensionsToRemoveForEmitPath(compilerOptions));
         }
         else {
-            emitOutputFilePathWithoutExtension = removeFileExtension(sourceFile.fileName);
+            emitOutputFilePathWithoutExtension = removeFileExtension(sourceFile.fileName, getExtensionsToRemoveForEmitPath(compilerOptions));
         }
 
         return emitOutputFilePathWithoutExtension + extension;
     }
 
+    export function getEmitFileNames(sourceFile: SourceFile, host: EmitHost) {

```

Comment:
Looking into this but closing this PR for now since I created separate one against master branch